### ISSUE TITLE
HaveIBeenPawned?

### ISSUE THREAD

**Issue by amaury1729** [reacherhq/check-if-email-exists#289]
Add a field `misc.have_i_been_pawned: true/false` which makes an API call to https://haveibeenpwned.com/

**Comment by NChechulin** [comment]
There is a small problem: haveibeenpwned's API costs $3.50/month.
Maybe consider scraping or a similar free API?

**Comment by amaurymartiny** [comment]
Ah, I wasn't aware it was paid. So maybe not, I don't think it's super high priority (and people can always make a separate API call for that).

I recall the author was open-sourcing it. Will it still be paid after?

**Comment by NChechulin** [comment]
On [API Key Page](https://haveibeenpwned.com/API/Key) they provide a link to a [blog post](https://www.troyhunt.com/authentication-and-the-have-i-been-pwned-api/), which says:

> Clearly not everyone will be happy with this so let me spend a bit of time here explaining the rationale. This fee is first and foremost to stop abuse of the API.

**Comment by DigitalGreyHat** [comment]
Hello I made my own API. It's free forever! And it works the same as haveibeenpwned.com. I try to make a PR soon.

**Comment by LeMoussel** [comment]
@DigitalGreyHat  Can you give some/more information about your API?

**Comment by olivermontes** [comment]
hi, any news? @DigitalGreyHat

**Comment by sylvain-reynaud** [comment]
Hello, I am currently working on this.

There are my thoughts:
* The easy way to use HIBP API is to pay for it,
* we can use a headless browser to bypass cloudflare protection to call https://haveibeenpwned.com/unifiedsearch/user@example.org

The problem with the cloudflare bypass is that we have to rely on a stealth browser. Otherwise cloudflare will be triggered. https://github.com/ultrafunkamsterdam/undetected-chromedriver seems to be the one with the biggest community. I did a PoC and the results are not reliable. It works ~70% of the time (30% of crash/no response). Another problem of the slealth browser is that it brings a lot of new dependencies with its maintainability need.

To my mind, implement the paid API is the way to go. Otherwise we can find another reliable and free API.

**Comment by amaurym** [comment]
> Otherwise we can find another reliable and free API.

Do people know of other free APIs? Ideally open-source. We can always add `misc.<other_api> = true/false`, and make those extra API calls configurable.

**Comment by sylvain-reynaud** [comment]
According to https://github.com/khast3x/h8mail#apis there are 3 free(ium) apis:
* https://breachdirectory.org/ (50 / month free)
* https://emailrep.io/ (250 free queries per month, up to 10 queries/day)
* https://hunter.io/pricing (25 free monthly searches)

**Comment by LeMoussel** [comment]
For information, I use[ Fingerprint Suite](https://apify.github.io/fingerprint-suite/) with [Playwright](https://playwright.dev/).
It's OK with [Antibot](https://bot.sannysoft.com/). I didn't test with Cloudflare.

**Comment by LeMoussel** [comment]
It seems OK in Firefox headless mode with this:
```python
import path from 'path';
import { fileURLToPath } from 'url';

import { firefox } from 'playwright';
import { FingerprintGenerator } from 'fingerprint-generator';
import { FingerprintInjector } from 'fingerprint-injector';

(async () => {
    const fingerprintGenerator = new FingerprintGenerator();

    const browserFingerprintWithHeaders = fingerprintGenerator.getFingerprint({
        devices: ['desktop'],
        browsers: ['firefox'],
    });

    const fingerprintInjector = new FingerprintInjector();
    const { fingerprint } = browserFingerprintWithHeaders;

    const browser = await firefox.launch({
        headless: true
    });

    // With certain properties, we need to inject the props into the context initialization
    const context = await browser.newContext({
        userAgent: fingerprint.userAgent,
        locale: fingerprint.navigator.language,
        viewport: fingerprint.screen,
    });

    // Attach the rest of the fingerprint
    await fingerprintInjector.attachFingerprintToPlaywright(context, browserFingerprintWithHeaders);

    const page = await context.newPage();

    await page.goto('https://haveibeenpwned.com/unifiedsearch/user@example.org');

    await page.screenshot({ path: path.join(path.dirname(fileURLToPath(import.meta.url)), 'playwright_test_headless.png') });

    await browser.close()
})();
```

**Comment by LeMoussel** [comment]
Yep! It's OK with [got-scraping](https://github.com/apify/got-scraping) 
`got-scraping` library has usually better success than other libraries due to header generation, http2 and browser ciphers. 

```javascript
import { gotScraping } from 'got-scraping';

(async () => {
    const response = await gotScraping({
        url: 'https://haveibeenpwned.com/unifiedsearch/user@example.org',
        headerGeneratorOptions:{
            browsers: ['firefox'],
            devices: ['desktop'],
        }
    });
    console.log(response.body)
})();
```

**Comment by sylvain-reynaud** [comment]
@LeMoussel wow I didn't know about this package, thank's :100: 

So I'm working on adding the feature by calling this URL https://haveibeenpwned.com/unifiedsearch/user@example.org

**Comment by sylvain-reynaud** [comment]
Hello, my PR is ready to be reviewed :)

**Comment by beshoo** [comment]
What's is this API?

**Comment by sylvain-reynaud** [comment]
I fixed the format and removed code that might break if a field is added on the API Response.

@beshoo it uses the haveibeenpwned API. The endpoint used is the one used by the front-end haveibeenpwned.com.

**Comment by amaurym** [comment]
The node.js libraries are probably more battle-tested, but I would like to keep this repo as pure Rust.

Also, I'm reluctant to use a headless browser for HIBP. It seems there's a risk that it'll become flaky/blocked one day, and the maintenance burden will likely fall on me. I propose to start with the paid API, as descrbied in https://github.com/reacherhq/check-if-email-exists/issues/289#issuecomment-1346149442. I'll gladly purchase the paid API and make it available on https://reacher.email 's SAAS plan.

**Comment by amaury1729** [comment]
Implemented in #1253, closing, thanks @sylvain-reynaud
